src/BlazorWebView/src/Maui/Android/WebKitWebViewClient.cs

```diff
@@ -38,24 +37,39 @@ public override bool ShouldOverrideUrlLoading(AWebView? view, IWebResourceReques
 			var requestUri = request?.Url?.ToString();
 			if (Uri.TryCreate(requestUri, UriKind.RelativeOrAbsolute, out var uri))
 			{
-				if (uri.Host == BlazorWebView.AppHostAddress &&
-					view is not null && 
-					request is not null && 
-					request.IsRedirect && 
-					request.IsForMainFrame)
+				if (AppOriginUri.IsBaseOfPage(uri))
```

Comment:
I'm not sure why we're using `IsBaseOfPage` here instead of `IsBaseOf`. Does it matter whether or not it's a page?

I thought that distinction was only relevant when deciding whether to do fallback routing.

Reply:
I guess my reasoning was that if we're trying to load a file, it seems like we should give the developer the option to open it in an external app/browser (i.e., we only load URLs in the view without invoking the `ExternalNavigationStarting` callback if they refer to other pages in the app). I could be making a wrong assumption here so I'm certainly happy to change it.

Reply:
> would intuitively consider static content in wwwroot to be internal

Do we have webview internal PDF support for instance across all the platforms?

Reply:
> (or perhaps routing patterns like /users/{fullname} matching /users/bob.gif - Bob Gif is a possible real name).

The developer could still support this if they wanted by adding their own handler for `ExternalNavigationStarting` and indicating that they want the URL to be opened in the webview (but this does seem like a bit of a hack IMO). Whereas, if we always loaded internal resources in the webview, I don't know that we have a way for the developer to override that behavior to support external navigation.

But there's something else: I think the existing fallback routing logic could still cause problems anyway for URLs ending in `.<something>`. I wonder if we should do something like re-processing the request as a navigation event if we fail to load an internal resource from the request URL (basically falling back to the host page if our initial guess was wrong). That way, users hopefully won't ever get the weird "There was no content at _____" page.

Reply:
> But there's something else: I think the existing fallback routing logic could still cause problems anyway for URLs ending in .<something>. I wonder if we should do something like re-processing the request as a navigation event if we fail to load an internal resource from the request URL (basically falling back to the host page if our initial guess was wrong). That way, users hopefully won't ever get the weird "There was no content at _____" page.

Hmm, it's unfortunate that Android's webview doesn't tell us the reason for the URL load event. WebView2 tells us whether or not it's a navigation, so we know unambiguously whether to use fallback logic or whether to regard it as a 404 if there's no match.

On Blazor WebAssembly/Server, we already have a strategy for dealing with not knowing whether it's a navigation: 

 * If a request matches some existing static content, we serve it
 * Otherwise, if it looks like a non-file (has no potential filename extension in the URL) we fall back on the index page
 * Otherwise, 404

It's not a perfect strategy as `/users/bob.gif` won't get the fallback when maybe you want it to (if that's meant to match `/users/{fullname}`). However people can customize this by editing their `Program.cs`. In practice, very few people have ever hit problems with this and complained about it.

> I think the big remaining question here is how we want to handle URLs that look like they point to internal files, but actually refer to pages. Do you have any thoughts on this @SteveSandersonMS @TanayParikh?

(Merging the threads as these perhaps come down to the same question)

I'm definitely happy to give developers control over this. In fact I expect we'll *have to*, because otherwise they have no way to customize this equivalent to what's available for Blazor WebAssembly/Server (i.e., changing the routing config).

The bit I was most thrown off by was referring to loading `/content/file.pdf` as **external** navigation when it's clearly internal to the base href.

Is it possible that the underlying problem here is a matter of terminology? We have this event called `ExternalNavigationStarting`. What if we just passed the whole question on to the app developer by generalizing this to be about all navigations, both internal and external? If we fired an event on *every* URL load, we could preconfigure some `UrlLoadingEventArgs` with a property called `UrlLoadingStrategy` set to whatever our basic guess is (e.g., load internally if an only if it's within the base href - and perhaps in the WebView2 case we take account of whether it's a navigation event since we have that info), then developers can add a handler to customize the logic any way they like for both internal and external loads. Developers who want to override the behavior for a specific subset of URL patterns can match those patterns in their handler, and for all other URLs, leave the preconfigured behavior unchanged.

Does this make sense? Am I up-to-date with all this, or was this already considered before?

Reply:
@SteveSandersonMS I'm definitely a fan of the idea of not making a hard distinction in our API between internal and external navigation events. If the rest of the team agrees that it's a good approach, I'm happy to start exploring that route.

One thing to note is that it doesn't seem `ShouldOverrideUrlLoading()` gets invoked for all URLs. I think if it's an internal URL that looks like a page, the method does not get invoked (which I suppose does mean that it is in fact unnecessary to call `AppOriginUri.IsBaseOfPage()`). I've also found that some links with handlers in other apps don't get run through `ShouldOverrideUrlLoading()`, e.g.:
```html
<a href="geo:124.028582,-29.201930" target="_blank">Click here for map</a>
```
The rules around this don't seem to be very clearly [documented](https://developer.android.com/reference/android/webkit/WebViewClient#shouldOverrideUrlLoading(android.webkit.WebView,%20android.webkit.WebResourceRequest)) either.

So, if developers are expecting that they'll be able to modify the way _any_ URL is handled, they could be mistaken. That said, if we made it clear that the event only applies to URLs that could conceivably refer to either a page or a file, then what I've just pointed out is probably a non-issue. We'll just have to make sure that the rules about what URLs can be overridden are consistent enough between each platform for our purposes.

Reply:
I'm guessing that those WebView events fire if the JS code in the app didn't already handle it. So presumably Blazor's JS code already handled the navigation so there's nothing for the WebView to do.

Reply:
> If we're planning on handling this as an "internal" navigation attempt, we'll also have to confirm that PDFs are supported on webviews across all platforms. Not a big deal, but just calling it out.

Do you think it would be fine to just load all "internal" URLs in the webview by default? If the developer can override this behavior, it seems to me like it would be overreaching a bit to decide which file types we should and shouldn't load in the webview by default (but I could be wrong about this).

Reply:
> Do you think it would be fine to just load all "internal" URLs in the webview by default?

Sorry I may have missed something here. Isn't that already our current behavior? Are you suggesting we don't go the route of `UrlLoadingStrategy` and just keep things as is?

>  the developer can override this behavior, it seems to me like it would be overreaching a bit to decide which file types we should and shouldn't load in the webview by default

Normally I'd agree, we want to minimize the complexity however possible. But opening PDFs in the device native PDF viewer seems like a fairly "core" feature. From the user perspective, I much rather have PDFs in the device native app as it's usually a much smoother experience and offers more functionality (like printing, saving to storage, sharing, etc) which may or may not work when it's within the webview. Then again, this may not be something we want to pursue for v1, I'm open to alternatives.

Reply:
@TanayParikh Sorry I didn't see your latest comment before I posted mine. Hence it may sound as if I was ignoring you!

> From the user perspective, I much rather have PDFs in the device native app as it's usually a much smoother experience and offers more functionality (like printing, saving to storage, sharing, etc)

I'm personally ambivalent about whether or not it's good UX to open the PDF in the WebView (one drawback is that it will lose any state held transiently in the hierarchy of Blazor components). I agree it's good that it should be *possible* for this to happen, and for developers to have means of controlling whether or not it does. However it might be too far for us to guarantee cross-platform support for this as it depends on what the platform's native webview happens to do if you point it at a PDF.

Reply:
> Instead, use a button or similar with an @onclick handler that does something useful for the target file (for example, opening it in an external app using MAUI APIs)

Ah I wasn't aware we had MAUI APIs to force the loading into an external (PDF Reader) app. In that case I agree this is all fine given developers at least have the optionality to open externally should it be something they want to do.

Reply:
> Sorry I may have missed something here. Isn't that already our current behavior? Are you suggesting we don't go the route of UrlLoadingStrategy and just keep things as is?

I may not have been clear, my bad. I was suggesting that the current default behavior for loading file-like URLs internal to the base href is fine, but we would give developers the option to override it via `UrlLoadingStrategy`.

> Normally I'd agree, we want to minimize the complexity however possible.

I guess my concerns with this were that:

1. If we special-case PDF handling, what should we do about other [image formats](https://en.wikipedia.org/wiki/Comparison_of_web_browsers#Image_format_support) (or other file types that browsers provide a preview for)? Are we going to evaluate their support on each platform and decide what best the default behavior is?
2. Would we ever change the default behavior if a webview on some platform adopted a sufficient built-in PDF viewer? For example, the Android webview doesn't have support for previewing PDFs (so presumably we would open in an external app), but iOS _does_  support it. So should we open within the webview on iOS but not Android? But then what if Android someday gets its own PDF viewer, will we change the default behavior again? Maybe we decide that for now we always open PDFs externally on both platforms, but what if the webview PDF viewers on both platforms improve to the extent where it's unnecessary to do so?

That said, maybe I'm overthinking this and it's fine to only consider the PDF format for now, and always open it externally on mobile platforms. I'm fine taking that route, just wanted to share some potential concerns 🙂

Reply:
Oops, I didn't refresh and missed a bunch of comments before posting, sorry!

Reply:
I should be unblocked for now, thanks @SteveSandersonMS! I think most of the work is done, the primary remaining TODOs are:
1. Completing/verifying the new link handling functionality on iOS.
2. Determining what should be done about pages with URLs that look like file names. I've implemented a solution for WinUI, but I'm not sure if we want to include it in this PR (we would probably need something for the other platforms as well).

Hoping to at least get **1.** finished soon so this PR can get reviewed/merged by EOD tomorrow.